compiler/rustc_mir/src/interpret/intern.rs

```diff
@@ -285,11 +286,13 @@ pub enum InternKind {
 /// tracks where in the value we are and thus can show much better error messages.
 /// Any errors here would anyway be turned into `const_err` lints, whereas validation failures
 /// are hard errors.
+#[tracing::instrument(skip(ecx))]
 pub fn intern_const_alloc_recursive<M: CompileTimeMachine<'mir, 'tcx>>(
     ecx: &mut InterpCx<'mir, 'tcx, M>,
     intern_kind: InternKind,
     ret: MPlaceTy<'tcx>,
-) where
+) -> Result<(), ErrorReported>
```

Comment:
Hm, I had quite deliberately made interning infallible a while ago.^^ Is this really necessary or just yet another hack to work around https://github.com/rust-lang/rust/issues/76064 ?

Reply:
well... infallible or not, it was reporting errors, so it wasn't really infallible, it just didn't report its failures.

The problem isn't #76064 in this case, even if the repro test would also be fixed by a fix for #76064 . This issue also occurs for

```rust
const FOO: *const u32 = {
    let x = 42;
    &x
};
```

which typechecks just fine. Basically any dangling pointer was causing this.

As you noted in https://github.com/rust-lang/rust/issues/78655#issuecomment-720403662 that error should stop compilation. We could change the interning error to a `delay_span_bug` and have it get reported by validation instead, but then validation needs to have a look at relocations within padding, which it currently does not.

Reply:
> well... infallible or not, it was reporting errors, so it wasn't really infallible, it just didn't report its failures.

That's fair.

> Basically any dangling pointer was causing this.

It was causing an error, but usually not an ICE -- IIRC we even had a testcase for that.

But I guess what I really care about is that interning does not return an `InterpError`, and that is still the case. Actually telling the outside world whether we did `span_err` seems reasonable and seems to be a more general pattern.

So, LGTM.

Reply:
>  then validation needs to have a look at relocations within padding, which it currently does not.

Interning is in a much better place to do this, so unless we really want the error to point to where in the value the problem occurs, I think the current approach works better.